### PR TITLE
ZCS code migration

### PR DESCRIPTION
**Pull request by kuangdai**
This PR creates module `deepxde.zcs` to support ZCS on tensorflow, pytorch and paddle backends (tf.compact.v1 not supported). Codes are tested on all examples and on both CPU and GPU with outstanding performance improvements confirmed. Examples and docs will be added in the next PR, as discusssed.

Non-local changes only include updating the top-level `__init__.py`.

### CODE REVIEW

**Comment by lululxvi** [comment]
Format the code using black https://github.com/psf/black

**Comment by kuangdai** [comment]
I have done reformatting with `black` default settings. Seems that this is not fully compatible with PEP 8 which I was using in PyCharm. But fine.

File: deepxde/__init__.py [hunk 1]
```diff
@@ -44,3 +45,6 @@
 )
 
 maps = nn
+
+# zcs
+from . import zcs
```
Comment (lululxvi) [inline deepxde/__init__.py:50]:
```
...
from . import utils
from . import zcs
```

File: deepxde/zcs/gradient.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
+"""
```
Comment (lululxvi) [inline deepxde/zcs/gradient.py:1]:
```
"""Gradients for ZCS"""
```

File: deepxde/zcs/gradient.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Gradients for ZCS with lazy evaluation.
+    """
+
+    def __init__(self, zcs_parameters, u):
+        self.zcs_parameters = zcs_parameters
+        self.n_dims = len(zcs_parameters["leaves"])
+
+        # create tensor $a_{ij}$
+        if backend_name == "pytorch":
```
Comment (lululxvi) [inline deepxde/zcs/gradient.py:21]: Reorder: tensorflow in the first, then pytorch, then paddle

File: deepxde/zcs/model.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
+"""
```
Comment (lululxvi) [inline deepxde/zcs/model.py:1]:
```
"""Model extension for ZCS support"""
```

File: deepxde/zcs/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from .gradient import LazyGrad
```
Comment (lululxvi) [inline deepxde/zcs/__init__.py:1]: Add `__all__`

File: deepxde/zcs/gradient.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            with self.a_tape:  # z_tape is already watching
+                return self.zcs_parameters["tape"].gradient(y, z)
+        elif backend_name == "pytorch":
+            return torch.autograd.grad(y, z, create_graph=True)[0]
+        elif backend_name == "paddle":
+            return paddle.grad(y, z, create_graph=True)[0]  # noqa
+        else:
+            raise NotImplementedError(
+                f"ZCS is not implemented for backend {backend_name}"
+            )
```
Comment (lululxvi) [inline deepxde/zcs/gradient.py:59]:
```
if backend_name == "pytorch":
    return ...
if backend_name == "paddle":
    return ...
raise NotImplementedError(...
```

File: deepxde/zcs/gradient.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            # no need to watch here because we don't need higher-orders w.r.t. a
+            return self.a_tape.gradient(y, self.a)
+        elif backend_name == "pytorch":
+            return torch.autograd.grad(y, self.a, create_graph=True)[0]
+        elif backend_name == "paddle":
+            return paddle.grad(y, self.a, create_graph=True)[0]  # noqa
+        else:
+            raise NotImplementedError(
+                f"ZCS is not implemented for backend {backend_name}"
+            )
```
Comment (lululxvi) [inline deepxde/zcs/gradient.py:72]: The same as previous one. Change `elif` to `if`

File: deepxde/zcs/model.py [hunk 8]
```diff
@@ -0,0 +1,307 @@
+"""Model extension for ZCS support"""
+
+from ..model import Model as BaseModel
+from .. import gradients as grad
+from ..backend import tf, torch, paddle  # noqa
```
Comment (lululxvi) [inline deepxde/zcs/model.py:5]:
```
from .. import gradients as grad
from ..backend import tf, torch, paddle # noqa
from ..model import Model as BaseModel
```

File: deepxde/zcs/__init__.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+__all__ = [
```
Comment (lululxvi) [inline deepxde/zcs/__init__.py:1]: """Compute gradients using Zero Coordinate Shift.

Reference: https://arxiv.org/abs/2311.00860
"""
